8252374: Add a new factory method to concatenate a sequence of BodyPublisher instances into a single publisher.

dfuch opened the pull request:
Continuing the review with a PR...

8252374: Add a new factory method to concatenate a sequence
         of BodyPublisher instances into a single publisher.
https://bugs.openjdk.java.net/browse/JDK-8252374


Draft CSR:
https://bugs.openjdk.java.net/browse/JDK-8252382

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (3/3 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8252374](https://bugs.openjdk.java.net/browse/JDK-8252374): Add a new factory method to concatenate a sequence of BodyPublisher instances into a single publisher.


### Reviewers
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/57/head:pull/57`
`$ git checkout pull/57`

ChrisHegarty commented:
/csr

dfuch commented:
Hi Jaikiran - the test results came back green.

ChrisHegarty commented on src/java.net.http/share/classes/java/net/http/HttpRequest.java:
```diff
@@ -654,5 +656,37 @@ public static BodyPublisher ofByteArrays(Iterable<byte[]> iter) {
+         * <p> If the sequence is empty an {@linkplain #noBody() empty} publisher
+         * is returned. Otherwise each publisher is lazily subscribed to in turn,
+         * until all the body bytes are published, an error occurs, or the
+         * returned publisher's subscription is cancelled.
```
We should probably say what the publishers `contentLength` will return.

dfuch replied:
OK.

ChrisHegarty commented on src/java.net.http/share/classes/java/net/http/HttpRequest.java:
```diff
@@ -654,5 +656,37 @@ public static BodyPublisher ofByteArrays(Iterable<byte[]> iter) {
         }
+
+        /**
+         * Creates a {@code BodyPublisher} that publishes a request
```
Trivially, should we replace "Creates" with "Returns" to allow for the possibility of, say, a singleton empty publisher (rather than having to "create" a new empty each time)

dfuch replied:
Good point!

ChrisHegarty commented on src/java.net.http/share/classes/java/net/http/HttpRequest.java:
```diff
@@ -654,5 +656,37 @@ public static BodyPublisher ofByteArrays(Iterable<byte[]> iter) {
+         * publisher's subscription is {@linkplain Flow.Subscription#cancel()
+         * cancelled}, or an error occurs while publishing the bytes, not all
+         * publishers in the sequence may be subscribed to.
+         * A publisher may be subscribed several times in sequence if the
```
I think that this sentence can be moved out of the implNote and into normative text

dfuch replied:
OK.

ChrisHegarty commented on src/java.net.http/share/classes/java/net/http/HttpRequest.java:
```diff
@@ -654,5 +656,37 @@ public static BodyPublisher ofByteArrays(Iterable<byte[]> iter) {
+         * logically equivalent to the concatenation of all bytes published
+         * by each publisher in the sequence.
+         */
+        public static BodyPublisher concat(BodyPublisher... publishers) {
```
It's a choice whether to allow an empty publishers array, or an array of just 1, or to disallow (throw IllegalArgumentException), but that will depend on expected usage.

dfuch replied:
Yes - returning `noBody()` seems more natural - since varargs allows you to call `BodyPublishers.concat()` with no parameters. I toyed with the idea of defining the method as `BodyPublisher concat(BodyPublisher first, BodyPublisher... rest)` but that's very cumbersome for the case where you already have a `List<BodyPublishers>` or a `BodyPublisher[]` array.

ChrisHegarty commented:
I added myself as reviewer on the CSR.

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/RequestPublishers.java:
```diff
@@ -491,4 +496,165 @@ public final void subscribe(Flow.Subscriber<? super ByteBuffer> subscriber) {
+
+
+    public static BodyPublisher concat(BodyPublisher... publishers) {
+        if (publishers == null || publishers.length == 0) {
```
OK - will remove.

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/RequestPublishers.java:
```diff
@@ -491,4 +496,165 @@ public final void subscribe(Flow.Subscriber<? super ByteBuffer> subscriber) {
+        AggregateSubscription(List<BodyPublisher> bodies, Flow.Subscriber<? super ByteBuffer> subscriber) {
+            this.bodies = new ConcurrentLinkedQueue<>(bodies);
+            this.subscriber = subscriber;
+            scheduler = SequentialScheduler.synchronizedScheduler(this::run);
```
Scheduler is a final variable in `AggregateSubscription`. Fixed to:

`    this.scheduler = ....;`

(though the `this.` is not strictly needed in this third instantiation)

pavelrappo replied:
Daniel, I believe Chris' concern is the escape of `this` before the (`AggregateSubscription `) object is fully constructed.

dfuch replied:
Sorry - where does it escape? It doesn't escape anywhere. We have used this idiom everywhere where a sequential scheduler was needed.

pavelrappo replied:
Right here:
```
scheduler = SequentialScheduler.synchronizedScheduler(this::run);
                                                      ^
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/RequestPublishers.java:
```diff
@@ -491,4 +496,165 @@ public final void subscribe(Flow.Subscriber<? super ByteBuffer> subscriber) {
+        }
+
+        @Override
+        public void request(long n) {
```
That will be thrown by Demand::increase below.

dfuch commented:
@ChrisHegarty As we discussed I have fixed the handling of negative requests, and hooked the publishers created by the new concat method to the TCK tests. 

This made me realize that the `AggregateSubscription` should be checking its cancellation status more often to make it sure it's reported as soon as possible to the upstream publishers. Similarly I updated the `PullPublisher` to check its error state more often. The issue here is when the error/cancellation gets reported in the same thread in which the sequential scheduler is running, and when the upstream publisher is busily forwarding items until its demand is exhausted. If you had a single `request()` call with a high demand - then you do need to check the error/cancellation state before forwarding the next item - and not only after/before each `request()` call to the upstream publisher.

I also tweaked `Demand` to generate an `IllegalArgumentException` message that contains the (optional) message string expected by the TCK.

ChrisHegarty commented on src/java.net.http/share/classes/jdk/internal/net/http/PullPublisher.java:
```diff
@@ -102,7 +102,7 @@ protected void run() {
                 }
 
-                while (demand.tryDecrement() && !cancelled) {
+                while (demand.tryDecrement() && !cancelled && error == null) {
```
The changes to `PullPublisher` seem like a general bug fix unrelated to `concat`. Is that true? Does it make sense to separate them out, or are they only applicable to a very narrow boundary scenario?

dfuch commented:
/integrate
